hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcConnection.java

```diff
@@ -321,14 +319,22 @@ public void run(boolean cancelled) throws IOException {
         if (cancelled) {
           setCancelled(call);
         } else {
+          Channel channel = channelRef.get();
           if (channel == null) {
-            connect();
+            try {
+              channelRef.compareAndSet(null, connect());
+              channel = channelRef.get();
```

Comment:
You discarded the return value of `compareAndSet`. If there are multiple connection attempts at the same time, one of it is leaked.